### ISSUE TITLE
ValueError: Unknown format code 'f' for object of type 'str'

### ISSUE THREAD

**Issue by rlittman16** [Magdoll/cDNA_Cupcake#70]
When using collapse_isoforms_by_sam.py I have this error: 

Traceback (most recent call last):
  File "/u/home/r/rlittman/project-zarlab/miniconda3/envs/python2/bin/collapse_isoforms_by_sam.py", line 4, in <module>
    __import__('pkg_resources').run_script('cupcake==5.8', 'collapse_isoforms_by_sam.py')
  File "/u/home/r/rlittman/project-zarlab/miniconda3/envs/python2/lib/python2.7/site-packages/pkg_resources/__init__.py", line 666, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/u/home/r/rlittman/project-zarlab/miniconda3/envs/python2/lib/python2.7/site-packages/pkg_resources/__init__.py", line 1453, in run_script
    exec(code, namespace, namespace)
  File "/u/project/zarlab/rlittman/miniconda3/envs/python2/lib/python2.7/site-packages/cupcake-5.8-py2.7-linux-x86_64.egg-info/scripts/collapse_isoforms_by_sam.py", line 245, in <module>
    main(args)
  File "/u/project/zarlab/rlittman/miniconda3/envs/python2/lib/python2.7/site-packages/cupcake-5.8-py2.7-linux-x86_64.egg-info/scripts/collapse_isoforms_by_sam.py", line 190, in main
    for recs in iter: # recs is {'+': list of list of records, '-': list of list of records}
  File "/u/home/r/rlittman/project-zarlab/miniconda3/envs/python2/lib/python2.7/site-packages/cupcake/tofu/branch/branch_simple2.py", line 99, in iter_gmap_sam
    ignored_fout.write("{0}\tIdentity {1:.3f} too low.\n".format(r.qID, r.identity))
ValueError: Unknown format code 'f' for object of type 'str'

I was using samtools sort, and got this error then I used:
`sort -k 3,3 -k 4,4n input.sam > input.sorted.sam`
as recommended by a previous issue, but I still have the same error.

I commented on the other issue, but didn't realize I can't reopen it. Sorry for the double post.

Thank you in advanced!

**Comment by Magdoll** [comment]
Most often this comes from corrupt/incomplete SAM files. Is it possible to re-run alignment, sort, and try again?

If fails again, I will request for you to upload the data for me to test.

Thanks,
--Liz

**Comment by rlittman16** [comment]
I wouldn't be able to upload the data, but I added print statements into the code to try and figure out what is going on. The reason it is crashing is because the r.identity is 'None'. I'm not entirely sure why this would happen, but this read is also a secondary mapped read (flag is 256), maybe that has something to do with it?

**Comment by Magdoll** [comment]
Hi @rlittman16 ,
Yes that usually means the SAM file was not sorted properly because it was trying to process an unaligned record. Or sometimes a corrupt alignment. 

That's why I recommended re-aligning, making sure the process finishes, and sort again

--Liz

**Comment by Magdoll** [comment]
Closed until further comment.

**Comment by bio15anu** [comment]
Hello, I get this same error after trying to use collapse_isoforms_by_sam.py on the sam output from STARlong following the recommended parameters here: https://github.com/Magdoll/cDNA_Cupcake/wiki/Best-practice-for-aligning-Iso-Seq-to-reference-genome:-minimap2,-deSALT,-GMAP,-STAR,-BLAT#using-star-to-align-reads-to-a-genome

The alignment seems to have finished successfully, and I have tried sorting with samtools and the sort method commented above. I have also tried manually filtering out any unmapped reads in case any remained, and filtering out non-primary mapped reads. I still run into this error even after re-alignment.

Could it be something specific to SAM files in the format produced by STAR?
